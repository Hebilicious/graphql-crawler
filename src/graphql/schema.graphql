### This file was generated by Nexus Schema
### Do not make changes to this file directly


"""Return the entire crawling data."""
type CrawledData {
  """List of found urls."""
  crawledData: [String!]!

  """Crawl current status."""
  isCrawling: Boolean!

  """The base url."""
  url: String!
}

"""The return type for the crawler Status subscription."""
type CrawlerStatusData {
  """List of urls currently being crawled."""
  currentlyCrawled: [String!]!

  """Urls done crawling."""
  doneCrawled: Int!
}

"""Input parameters for the crawlUrl query"""
input CrawlUrlInput {
  """Delay for the crawler"""
  delay: Int = 0

  """The maximum amount of crawled URLs per crawl."""
  maxUrls: Int = 100

  """Base url for the crawler."""
  url: String
}

"""
A date string, such as 2007-12-03, compliant with the `full-date` format outlined in section 5.6 of the RFC 3339 profile of the ISO 8601 standard for representation of dates and times using the Gregorian calendar.
"""
scalar Date

"""
The `JSONObject` scalar type represents JSON objects as specified by [ECMA-404](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf).
"""
scalar JSONObject @specifiedBy(url: "http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf")

type Query {
  """Start a crawler for a given URL"""
  crawlUrl(input: CrawlUrlInput): String!

  """Generate a sitemap for a crawled URL."""
  generateSitemap(
    """
    Wether or not we want to include the Crawled externals links. A valid Sitemap XML shouldn't include them.
    """
    externalLinks: Boolean = false

    """
    This needs to be a crawled URL, ie crawlUrl needs to finish running first.
    """
    url: String!
  ): SiteMapData!

  """Return a paginated list of Crawled Data."""
  getCrawledData(
    """Skip the first n elements."""
    skip: Int = 0

    """Take n elements."""
    take: Int = 5
  ): [CrawledData!]!
  ok: Boolean!
}

"""Sitemap in XML and JSON."""
type SiteMapData {
  """Crawled Data JSON array."""
  json: JSONObject!

  """XML Sitemap string."""
  xml: String!
}

type Subscription {
  """
  Subscribtion endpoint that gives information about the ongoing crawling operations.
  """
  crawlerStatus: CrawlerStatusData!
}
