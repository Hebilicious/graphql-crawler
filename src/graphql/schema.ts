import { makeSchema, queryField } from "@nexus/schema"
import * as path from "path"

import { CrawlerStatus, CrawlUrl, GetCrawledData } from "./types/crawl"
import { GQLDate, GQLJson } from "./types/scalar"
import { GenerateSitemap } from "./types/sitemap"

/**
 * This is a demo query that we can use for health-checking.
 */
const Ok = queryField("ok", { type: "Boolean", resolve: () => true })

const types = [Ok, CrawlUrl, CrawlerStatus, GetCrawledData, GenerateSitemap, GQLDate, GQLJson]

export const schema = makeSchema({
    types,
    outputs: {
        schema: path.join(__dirname, "./schema.graphql"),
        typegen: path.join(__dirname, "../../node_modules/@types/nexus-typegen/index.d.ts")
    },
    typegenAutoConfig: {
        contextType: "Context.Context",
        sources: [
            {
                source: path.resolve(__dirname, "./context.ts"),
                alias: "Context"
            }
        ]
    }
})
