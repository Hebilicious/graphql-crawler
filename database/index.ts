interface DatabaseItem {
    isCrawling: boolean
    crawledData: url[]
}

export type url = string

interface insertData {
    url: url
    databaseItem: DatabaseItem
}

const datastore: Map<url, DatabaseItem> = new Map()

/* eslint-disable @typescript-eslint/no-unused-vars */
export const db = async () => {
    return {
        crawler: {
            get: async (url: url) => datastore.get(url),
            upsert: async ({ url, databaseItem }: insertData) => datastore.set(url, databaseItem),
            getAll: async () => [...datastore],
            getCurrentlyCrawled: async () => [...datastore].filter(([key, { isCrawling }]) => isCrawling),
            getDoneCrawled: async () => [...datastore].filter(([key, { isCrawling }]) => !isCrawling),
            has: async (url: url) => datastore.has(url),
            delete: async (url: url) => datastore.delete(url),
            flush: async () => datastore.clear()
        }
    }
}
