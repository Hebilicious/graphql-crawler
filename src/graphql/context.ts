import { PubSub } from "apollo-server-fastify"
import { db } from "../../database"
import { dispatch } from "../services/handleCrawl"

const pubsub = new PubSub()

export interface Context {
    pubsub: PubSub
    db: typeof db
    dispatch: typeof dispatch
}

export const createContext = (): Context => ({ pubsub, db, dispatch })
