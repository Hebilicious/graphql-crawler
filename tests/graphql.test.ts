import { ApolloServerTestClient, createTestClient } from "apollo-server-testing"
import { createContext } from "../src/graphql/context"
import { schema } from "../src/graphql/schema"

import { ApolloServer, gql } from "apollo-server-fastify"
import { db } from "../database"

let Client: ApolloServerTestClient

beforeAll(() => {
    const server = new ApolloServer({ schema, context: createContext })
    Client = createTestClient(server)
})

test("Can send the ok query", async () => {
    const { query } = Client
    const { data } = await query({
        query: gql`
            query okQuery {
                ok
            }
        `
    })
    if (!data) throw new Error("No data")
    expect(data).toHaveProperty("ok")
    expect(data.ok).toBe(true)
})

describe("Crawl", () => {
    test("Can use the crawlUrl operation", async () => {
        const { query } = Client
        const { data } = await query({
            query: gql`
                query crawl {
                    crawlUrl(input: { url: "https://emmanuel.style" })
                }
            `
        })
        if (!data) throw new Error("No data")
        expect(data).toHaveProperty("crawlUrl")
        expect(data.crawlUrl).toContain("emmanuel.style")
    })

    test("Can use the getCrawledData operation", async () => {
        const database = await db()
        //Insert 50 test items
        Array.from({ length: 50 }).forEach(async (n, index) => {
            await database.crawler.upsert({
                url: `test-${index}`,
                databaseItem: { crawledData: [], isCrawling: true }
            })
        })
        const { query } = Client
        const { data } = await query({
            query: gql`
                query crawlData {
                    getCrawledData(skip: 1, take: 12) {
                        url
                        isCrawling
                        crawledData
                    }
                }
            `
        })
        if (!data) throw new Error("No data")
        expect(data).toHaveProperty("getCrawledData")
        expect(Array.isArray(data.getCrawledData)).toBe(true)
        expect(data.getCrawledData.length).toBeGreaterThanOrEqual(12)
        //Remove the 50 test items
        Array.from({ length: 50 }).forEach(async (n, index) => {
            await database.crawler.delete(`test-${index}`)
        })
    })
})

describe("Sitemap", () => {
    beforeAll(async () => {
        const database = await db()
        await database.crawler.upsert({
            url: `https://example.com`,
            databaseItem: {
                crawledData: ["https://example.com/foo", "https://example.com/bar", "https://test.com/bar"],
                isCrawling: false
            }
        })
    })
    afterAll(async () => {
        const database = await db()
        await database.crawler.delete("https://example.com")
    })
    test("Can generate a valid Sitemap", async () => {
        const { query } = Client
        const { data } = await query({
            query: gql`
                query sitemap {
                    generateSitemap(url: "https://example.com", externalLinks: false) {
                        json
                        xml
                    }
                }
            `
        })
        if (!data) throw new Error("No data")
        expect(data).toHaveProperty("generateSitemap.json.hrefs")
        expect(data).toHaveProperty("generateSitemap.xml")
        expect(data.generateSitemap.xml).toMatchSnapshot()
        expect(data.generateSitemap.json.hrefs).toHaveLength(2)
    })

    test("Can generate a 'sitemap' with external links.", async () => {
        const { query } = Client
        const { data } = await query({
            query: gql`
                query sitemap {
                    generateSitemap(url: "https://example.com", externalLinks: true) {
                        json
                        xml
                    }
                }
            `
        })
        if (!data) throw new Error("No data")
        expect(data).toHaveProperty("generateSitemap.json.hrefs")
        expect(data).toHaveProperty("generateSitemap.xml")
        expect(data.generateSitemap.xml).toMatchSnapshot()
        expect(data.generateSitemap.json.hrefs).toHaveLength(3)
    })
})

describe("End to End", () => {
    beforeAll(async () => {
        const { query } = Client
        await query({
            query: gql`
                query crawl {
                    crawlUrl(input: { url: "https://google.com" })
                }
            `
        })
        await query({
            query: gql`
                query crawl {
                    crawlUrl(input: { url: "https://emmanuel.style" })
                }
            `
        })
    })
    test("Crawling an url correctly write to the database", async () => {
        const database = await db()
        const googleExists = await database.crawler.has("https://google.com")
        const emmanuelExists = await database.crawler.has("https://emmanuel.style")
        expect(googleExists).toBe(true)
        expect(emmanuelExists).toBe(true)
    })
})
