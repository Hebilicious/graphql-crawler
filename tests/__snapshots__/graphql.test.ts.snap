// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sitemap Can generate a 'sitemap' with external links. 1`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:xsi=\\"http://www.w3.org/2001/XMLSchema-instance\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\"><url><loc>https://example.com/foo</loc><changefreq>weekly</changefreq></url><url><loc>https://example.com/bar</loc><changefreq>weekly</changefreq></url><url><loc>https://test.com/bar</loc><changefreq>weekly</changefreq></url></urlset>"`;

exports[`Sitemap Can generate a valid Sitemap 1`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:xsi=\\"http://www.w3.org/2001/XMLSchema-instance\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\"><url><loc>https://example.com/foo</loc><changefreq>weekly</changefreq></url><url><loc>https://example.com/bar</loc><changefreq>weekly</changefreq></url></urlset>"`;
